projectile.el

```diff
@@ -3215,7 +3215,8 @@ This command will first prompt for the directory the file is in."
   (cl-mapcan
    (lambda (project)
      (when (file-exists-p project)
-       (let ((default-directory project))
+       (let ((default-directory project)
+             (projectile-cached-project-root nil))
```

Comment:
Perhaps this approach could also be used to switch project? That is, functions used to as switch project actions (`projectile-find-file` ...) could take the project root as optional parameter and `projectile-switch-project-by-name` could then simply do something like this:

```
(run-hooks 'projectile-before-switch-project-hook)
(funcall action project-to-switch)
(run-hooks 'projectile-after-switch-project-hook)
```

without needing to let-bind `default-directory`? I'm not sure how this would affect the caching mechanism for project roots and names or if it may not be suitable for all potential actions.

Speaking about `projectile-switch-project-by-name`, I think the `with-temp-buffer` call inside it is useless now that that the action is no longer called from the temp buffer (commit 72a8be5 ). Unless I'm mistaken, it just creates a temp buffer, loads the project-to-siwtch's dir-local variables into this buffer and then immediately kills it.

And speaking about this commit, it reintroduced issue #1005 (for me at least) that had been solved by calling the action from the temp buffer (but that had created other issues). Perhaps this issue would be solved again if `projectile-switch-project-by-name`no longer let-binds `default-directory`?